File: src/main/java/graphql/nadel/ServiceExecutionParameters.java
package graphql.nadel;

import graphql.PublicApi;
import graphql.language.Document;
import graphql.language.FragmentDefinition;
import graphql.language.OperationDefinition;

import java.util.LinkedHashMap;
import java.util.Map;

import static graphql.Assert.assertNotNull;


@PublicApi
public class ServiceExecutionParameters {

    private final Document query;
    private final Object context;
    private final Map<String, Object> variables;
    private final Map<String, FragmentDefinition> fragments;
    private final OperationDefinition operationDefinition;

    private ServiceExecutionParameters(Document query, Object context, Map<String, Object> variables, Map<String, FragmentDefinition> fragments, OperationDefinition operationDefinition) {
        this.query = assertNotNull(query);
        this.variables = assertNotNull(variables);
        this.fragments = assertNotNull(fragments);
        this.operationDefinition = assertNotNull(operationDefinition);
        this.context = context;
    }

    public Document getQuery() {
        return query;
    }

    public Object getContext() {
        return context;
    }

    public Map<String, Object> getVariables() {
        return variables;
    }

    public Map<String, FragmentDefinition> getFragments() {
        return fragments;
    }

    public OperationDefinition getOperationDefinition() {
        return operationDefinition;
    }

    public OperationDefinition.Operation getOperation() {
        return operationDefinition.getOperation();
    }


    public static Builder newServiceExecutionParameters() {
        return new Builder();
    }

    public static class Builder {
        private Document query;
        private Object context;
        private Map<String, Object> variables = new LinkedHashMap<>();
        private Map<String, FragmentDefinition> fragments = new LinkedHashMap<>();
        private OperationDefinition operationDefinition;

        private Builder() {
        }

        public Builder query(Document query) {
            this.query = query;
            return this;
        }

        public Builder context(Object context) {
            this.context = context;
            return this;
        }

        public Builder variables(Map<String, Object> variables) {
            this.variables.putAll(variables);
            return this;
        }

        public Builder fragments(Map<String, FragmentDefinition> fragments) {
            this.fragments.putAll(fragments);
            return this;
        }

        public Builder operationDefinition(OperationDefinition operationDefinition) {
            this.operationDefinition = operationDefinition;
            return this;
        }

        public ServiceExecutionParameters build() {
            return new ServiceExecutionParameters(query, context, variables, fragments, operationDefinition);
        }
    }
}

File: lib/src/main/java/graphql/nadel/ServiceExecutionParameters.kt
package graphql.nadel

import graphql.GraphQLContext
import graphql.execution.ExecutionId
import graphql.introspection.Introspection.TypeNameMetaFieldDef
import graphql.language.Document
import graphql.language.OperationDefinition
import graphql.nadel.engine.NadelServiceExecutionContext
import graphql.normalized.ExecutableNormalizedField

class ServiceExecutionParameters internal constructor(
    val query: Document,
    val context: Any?,
    val graphQLContext: GraphQLContext,
    val variables: Map<String, Any>,
    val operationDefinition: OperationDefinition,
    val executionId: ExecutionId,
    val serviceExecutionContext: NadelServiceExecutionContext,
    /**
     * @return details abut this service hydration or null if it's not a hydration call
     */
    val hydrationDetails: ServiceExecutionHydrationDetails?,
    /**
     * All transformed underlying-schema root fields represented by [query], including artificial fields.
     */
    val executableNormalizedFields: List<ExecutableNormalizedField>,
) {
    /**
     * The first non-__typename underlying field, or the first field if all are __typename.
     * Retains the legacy selection behavior; additional service-call fields are omitted.
     */
    @Deprecated(
        message = "Use executableNormalizedFields and explicitly handle all fields in the service call.",
    )
    val executableNormalizedField: ExecutableNormalizedField =
        executableNormalizedFields.firstOrNull { it.fieldName != TypeNameMetaFieldDef.name }
            ?: executableNormalizedFields.first()

    val isHydrationCall: Boolean
        get() = hydrationDetails != null
}
